pandas/tests/test_strings.py

```diff
@@ -2760,6 +2761,138 @@ def test_str_cat_raises_intuitive_error(self):
         with tm.assert_raises_regex(ValueError, message):
             s.str.cat('    ')
 
+    def test_str_cat_align(self):
+        # https://github.com/pandas-dev/pandas/issues/18657
+        r = Series(['a', 'b', 'c', 'd'])
+        t = Series(['d', 'a', 'e', 'b'], index=[3, 0, 4, 1])
+        x = Series(['f', 'f', 'g', 'f'], index=[0, 2, 3, 4],
+                   dtype='category')
+        d = concat([t, t], axis=1)
+
+        # check FutureWarning and that join=None defaults to no alignment
+        exp = Series(['ad', 'ba', 'ce', 'db'])
+        with tm.assert_produces_warning(expected_warning=FutureWarning):
+            tm.assert_series_equal(r.str.cat(t), exp)
+
+        # aligned with different joins
+        # left
+        exp = Series(['aa', 'bb', np.nan, 'dd'])
```

Comment:
this needs to be parametrized, potentially splitting into multiple tests. you have added a ton of logic and its not really clear whether all paths are tested.

Reply:
Can you give me an example how to do the parametrisation you're talking about?

Reply:
Diff coverage is 100%, BTW

Reply:
The newest commit contains compartmentalized and parametrized tests - could you comment if this is the way you imagined it. Regrouping them in other ways (or changing parameters) is not an issue.